src/test/crimson/test_messenger.cc

```diff
@@ -19,175 +29,339 @@ static seastar::future<> test_echo(unsigned rounds,
+                    if (keepalive_dist(rng)) {
+                      count_keepalive += 1;
+                      return conn->keepalive()
+                        .then([&count_keepalive] {
+                          return seastar::make_ready_future<seastar::stop_iteration>(
+                            seastar::stop_iteration::no);
+                        });
+                    } else {
+                      count_ping += 1;
+                      return conn->send(MessageRef{new MPing(), false})
+                        .then([] {
+                          return seastar::make_ready_future<seastar::stop_iteration>(
+                            seastar::stop_iteration::yes);
+                        });
+                    }
+                  });
+              }).then([this, conn] {
+                auto found = pending_conns.find(conn);
+                if (found == pending_conns.end())
+                  throw std::runtime_error{"Not connected."};
+                return found->second.get_future();
               });
-            }).finally([&] {
-              if (verbose) {
-                std::cout << "client shutting down" << std::endl;
-              }
-              return t.client.messenger.shutdown();
-            });
-        }).finally([&] {
-          if (verbose) {
-            std::cout << "server shutting down" << std::endl;
-          }
-          return t.server.messenger.shutdown();
+          });
+      }
+    };
+  };
+
+  logger().info("test_echo():");
+  return seastar::when_all_succeed(
+      ceph::net::create_sharded<test_state::Server>(),
+      ceph::net::create_sharded<test_state::Server>(),
+      ceph::net::create_sharded<test_state::Client>(rounds, keepalive_ratio),
+      ceph::net::create_sharded<test_state::Client>(rounds, keepalive_ratio))
+    .then([rounds, keepalive_ratio](test_state::Server *server1,
+                                    test_state::Server *server2,
+                                    test_state::Client *client1,
+                                    test_state::Client *client2) {
+      // start servers and clients
+      entity_addr_t addr1;
+      addr1.parse("127.0.0.1:9010", nullptr);
+      addr1.set_type(entity_addr_t::TYPE_LEGACY);
+      entity_addr_t addr2;
+      addr2.parse("127.0.0.1:9011", nullptr);
+      addr2.set_type(entity_addr_t::TYPE_LEGACY);
+      return seastar::when_all_succeed(
+          server1->init(entity_name_t::OSD(0), "server1", 1, addr1),
+          server2->init(entity_name_t::OSD(1), "server2", 2, addr2),
+          client1->init(entity_name_t::OSD(2), "client1", 3),
+          client2->init(entity_name_t::OSD(3), "client2", 4))
+      // dispatch pingpoing
+        .then([client1, client2, server1, server2] {
+          return seastar::when_all_succeed(
+              client1->dispatch_pingpong(server1->msgr->get_myaddr(), true),
+              client1->dispatch_pingpong(server2->msgr->get_myaddr(), false),
```

Comment:
why do we need two tests for testing `!foreign_dispatch`:
```c++
client1->dispatch_pingpong(server2->msgr->get_myaddr(), false),
```
and 
```c++
client2->dispatch_pingpong(server1->msgr->get_myaddr(), false),
```
?
simply to see if a `SocketConnection` is able to handle i/o sent from different shards in parallel?

Reply:
Yes, the four `dispatch_pingpong(...)` combinations are to test connecting in parallel, accepting in parallel, and operating the connection reference from a foreign/local core.

Reply:
could accompany it with a comment to explain what's the intention of having multiple client here?

Reply:
comment added.